libsignal-service/src/cipher.rs

```diff
@@ -2,22 +2,27 @@ use crate::{
     content::{Content, Metadata},
     envelope::Envelope,
     push_service::ServiceError,
+    sender::OutgoingPushMessage,
     ServiceAddress,
 };
 
 use libsignal_protocol::{
+    messages::CiphertextType,
     messages::{PreKeySignalMessage, SignalMessage},
-    Address as ProtocolAddress, Context, Deserializable, SessionCipher,
-    StoreContext,
+    Address as ProtocolAddress, Context, Deserializable, Serializable,
+    SessionCipher, StoreContext,
 };
+
+use log::trace;
 use prost::Message;
 
 /// Decrypts incoming messages and encrypts outgoing messages.
 ///
 /// Equivalent of SignalServiceCipher in Java.
+#[derive(Clone, Debug)]
 pub struct ServiceCipher {
     store: StoreContext,
-    local: ServiceAddress,
+    pub local: ServiceAddress,
```

Comment:
`pub(crate)` seems more appropriate. I had personally changed the constructor of `MessageSender` to have another copy of the local address, but either way is okay.